### PR TITLE
Add get_badges_by_string method with pagination 

### LINKED ISSUE #267: Add get_badges_by_string() with DB-level pagination to TahrirDatabase
Adds a new `get_badges_by_string()` method to `TahrirDatabase` that searches 
badges by name, description and tags with proper DB-level pagination.

### PR DESCRIPTION
**Pull request by Daniel-1600**
solves #267

### CODE REVIEW

**Review by sdglitched — CHANGES_REQUESTED** [review]
Is this PR completed? I don't see test cases for the new function.

**Comment by Daniel-1600** [comment]
@sdglitched I had forgotten to add the tests but I have added them

**Comment by Daniel-1600** [comment]
what I did here was that the db logic was at the endpoint and when the endpoint for searching badhes by string was called it would fetch all the 400 badges which the would be sliced off by the python , I decided to shift that to api where the slicing would be done in the database level which is more efficient and also good for the memory servers . after shifting the logic to the api i also adedd pagination which resolves the issue of fetching alla the badges only to return a few for my pagination it fetches only a few badges 

Before the pagination on action it returns all the badges 

https://github.com/user-attachments/assets/fc0ddd35-523c-4f84-a6ca-9dd53f68cb2b

after pagination by having a limit of 1 it only returns one badge 

https://github.com/user-attachments/assets/71d8ce03-014f-41f7-8cf3-55fffc1771bc

**Comment by thedunncodes** [comment]
Your `castup` length is inaccurate it should match the page size requested. it'd be better to return the actual paginated size

**Review by sdglitched — CHANGES_REQUESTED** [review]
@Daniel-1600, again, I don't understand what is so difficult about following the 50-character limit of the commit message header?

Please refer to the comment in PR https://github.com/fedora-infra/tahrir-api/pull/263 and make changes.

Only then will the further review be made.

File: tests/test_dbapi.py [hunk 1]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+    api.add_badge(
+        name="OSS Contributor",
+        image="oss_contributor.png",
+        desc="Active contributor to open source projects and python ecosystems",
+        criteria="Contribute to 3 open source projects",
+        issuer_id=dummy_issuer_id,
+        tags="opensource, python, community",
+    )
+
+    return 5
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Move this fixture to the top after the rest of the existing fixtures.

File: tests/test_dbapi.py [hunk 2]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+
+
+def test_search_by_name_exact_match(api, search_test_badges):
+    """Test searching for a badge by exact name match."""
+    result = api.get_badges_by_string("Python Expert")
+
+    assert result["total"] >= 1
+    assert len(result["badges"]) >= 1
```
Comment (sdglitched) [inline tests/test_dbapi.py]: There is a single `Python Expert` badge, so why are you asserting it as >= 1?

File: tests/test_dbapi.py [hunk 3]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+    mixed_ids = {b.id for b in result_mixed["badges"]}
+
+    assert lower_ids == upper_ids == mixed_ids
+
+
+def test_search_by_description(api, search_test_badges):
+    """Test searching for a badge by description content."""
+    result = api.get_badges_by_string("web developer")
+
+    assert result["total"] >= 1
```
Comment (sdglitched) [inline tests/test_dbapi.py]: There is a single badge with 'web developer' in its description, so why are you asserting it as >= 1?

File: tests/test_dbapi.py [hunk 4]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+    # Should find "Frontend Master" whose description contains "web developer"
+    badge_names = [b.name for b in result["badges"]]
+    assert "Frontend Master" in badge_names
+
+
+def test_search_by_tags(api, search_test_badges):
+    """Test searching for a badge by tags."""
+    result = api.get_badges_by_string("documentation")
+
+    assert result["total"] >= 1
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Again, a single badge with the tag 'documentation'.

File: tests/test_dbapi.py [hunk 5]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+    )
+
+    result = api.get_badges_by_string("C++")
+
+    assert result["total"] >= 1
+    badge_names = [b.name for b in result["badges"]]
+    assert "C++ Expert" in badge_names
+
+
+def test_search_with_spaces(api, search_test_badges):
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Search with space is already tested in the first test case. Remove this one.

File: tests/test_dbapi.py [hunk 6]
```diff
@@ -675,3 +675,249 @@ def test_get_all_series_empty(api):
...
+        image="cpp_expert.png",
+        desc="Expert in C++ programming",
+        criteria="Complete C++ projects",
+        issuer_id=dummy_issuer_id,
+        tags="cpp, programming",
+    )
+
+    result = api.get_badges_by_string("C++")
+
+    assert result["total"] >= 1
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Again, a single badge with special characters.

**Review by sdglitched — CHANGES_REQUESTED** [review]
Good work with following the guidelines and coding standards.

Please make the requested change to the fixtrue's position, and I would want you to parameterise the tests into two parts. One for the general search, which will include all the edge cases for the searching string, and another one for the pagination.

File: tests/test_dbapi.py (hunk 2 above)
Comment (Daniel-1600) [reply tests/test_dbapi.py]: ooh lemme change that it should be assert result["total"] == 1
assert len(result["badges"]) == 1

File: tests/test_dbapi.py (hunk 3 above)
Comment (Daniel-1600) [reply tests/test_dbapi.py]: I've seen its wrong ill rectify that

**Comment by sdglitched** [comment]
@Daniel-1600, reiterating on my comment from your other PR https://github.com/fedora-infra/tahrir-api/pull/265#issuecomment-4221046313

Do you think the the first commit message is providing enough details?

Adding function name in commit headers leads to having longer length. To utilise the limited number of characters better, having the intent first makes more sense. Then further details can be added in the description.

Let me give you an example:
```shell
feat: add badge search functionality 

Create get_badges_by_string to filter badges by
name, description and tags.
Add pagination support using offset and limit.

Assisted-by: Claude
Signed-off-by: Daniel Mungai <chegedan699@gmail.com>
```

Also, while looking at the test cases again, I can say it can be easily parameterized. Please check how to parameterize test cases and then parameterize your tests in two test cases, one test for different filter condition and other one for different pagination condition.

**Comment by Daniel-1600** [comment]
> @Daniel-1600, reiterating on my comment from your other PR [#265 (comment)](https://github.com/fedora-infra/tahrir-api/pull/265#issuecomment-4221046313)
> 
> Do you think the the first commit message is providing enough details?
> 
> Adding function name in commit headers leads to having longer length. To utilise the limited number of characters better, having the intent first makes more sense. Then further details can be added in the description.
> 
> Let me give you an example:
> 
> ```shell
> feat: add badge search functionality 
> 
> Create get_badges_by_string to filter badges by
> name, description and tags.
> Add pagination support using offset and limit.
> 
> Assisted-by: Claude
> Signed-off-by: Daniel Mungai <chegedan699@gmail.com>
> ```
> 
> Also, while looking at the test cases again, I can say it can be easily parameterized. Please check how to parameterize test cases and then parameterize your tests in two test cases, one test for different filter condition and other one for different pagination condition.

for the explanation won't it add up the charaters or its is not considered?

**Comment by Daniel-1600** [comment]
> @Daniel-1600, reiterating on my comment from your other PR [#265 (comment)](https://github.com/fedora-infra/tahrir-api/pull/265#issuecomment-4221046313)
> 
> Do you think the the first commit message is providing enough details?
> 
> Adding function name in commit headers leads to having longer length. To utilise the limited number of characters better, having the intent first makes more sense. Then further details can be added in the description.
> 
> Let me give you an example:
> 
> ```shell
> feat: add badge search functionality 
> 
> Create get_badges_by_string to filter badges by
> name, description and tags.
> Add pagination support using offset and limit.
> 
> Assisted-by: Claude
> Signed-off-by: Daniel Mungai <chegedan699@gmail.com>
> ```
> 
> Also, while looking at the test cases again, I can say it can be easily parameterized. Please check how to parameterize test cases and then parameterize your tests in two test cases, one test for different filter condition and other one for different pagination condition.

now I get it I thought adding the file change gives more context but now i now understand

**Comment by Daniel-1600** [comment]
> @Daniel-1600, reiterating on my comment from your other PR [#265 (comment)](https://github.com/fedora-infra/tahrir-api/pull/265#issuecomment-4221046313)
> 
> Do you think the the first commit message is providing enough details?
> 
> Adding function name in commit headers leads to having longer length. To utilise the limited number of characters better, having the intent first makes more sense. Then further details can be added in the description.
> 
> Let me give you an example:
> 
> ```shell
> feat: add badge search functionality 
> 
> Create get_badges_by_string to filter badges by
> name, description and tags.
> Add pagination support using offset and limit.
> 
> Assisted-by: Claude
> Signed-off-by: Daniel Mungai <chegedan699@gmail.com>
> ```
> 
> Also, while looking at the test cases again, I can say it can be easily parameterized. Please check how to parameterize test cases and then parameterize your tests in two test cases, one test for different filter condition and other one for different pagination condition.

ooh and also for the commit message should include changes i am making ie if its a feature ?

**Comment by Daniel-1600** [comment]
> @Daniel-1600, reiterating on my comment from your other PR [#265 (comment)](https://github.com/fedora-infra/tahrir-api/pull/265#issuecomment-4221046313)
> 
> Do you think the the first commit message is providing enough details?
> 
> Adding function name in commit headers leads to having longer length. To utilise the limited number of characters better, having the intent first makes more sense. Then further details can be added in the description.
> 
> Let me give you an example:
> 
> ```shell
> feat: add badge search functionality 
> 
> Create get_badges_by_string to filter badges by
> name, description and tags.
> Add pagination support using offset and limit.
> 
> Assisted-by: Claude
> Signed-off-by: Daniel Mungai <chegedan699@gmail.com>
> ```
> 
> Also, while looking at the test cases again, I can say it can be easily parameterized. Please check how to parameterize test cases and then parameterize your tests in two test cases, one test for different filter condition and other one for different pagination condition.

It actually doesn't , i will definitely rectify on that ,thanks for the feedback.

**Comment by sdglitched** [comment]
> for the explanation won't it add up the charaters or its is not considered?

So, the basic commit convention says that the header should have at most 50 characters, and the description is allowed to have 72 characters per line, i.e., there can be multiple lines in the description, but each line should not exceed 72 characters.
image

> ooh and also for the commit message should include changes i am making ie if its a feature ?

It is good to have, not a hard and fast rule.

You can refer [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) for more details, but keep in mind its examples do not follow the 50/72 suggestion.

File: tests/test_dbapi.py [hunk 7]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+@pytest.mark.parametrize(
+    "query,expected_total,must_contain",
+    [
+        ("Python Expert", 1, ["Python Expert"]),
+        ("python", 2, ["Python Expert", "OSS Contributor"]),
+        ("PYTHON", 2, ["Python Expert", "OSS Contributor"]),
+        ("web developer", 1, ["Frontend Master"]),
+        ("documentation", 1, ["Doc Writer"]),
+        ("nonexistent_badge_xyz", 0, []),
+        ("", 5, []),
```
Comment (sdglitched) [inline tests/test_dbapi.py]: This test expects `>= 5` results, but doesn't validate that it returns exactly the 5 test fixtures. Please add them.

File: tests/test_dbapi.py [hunk 8]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+    ],
+)
+def test_search_filter_conditions(query, expected_total, must_contain, api, search_test_badges):
+    result = api.get_badges_by_string(query)
+
+    assert result["total"] >= expected_total
```
Comment (sdglitched) [inline tests/test_dbapi.py]: The result should be the same as that of the expected total. Please change this and make the necessary change in test parameters if needed.
```suggestion
    assert result["total"] = expected_total
```

File: tests/test_dbapi.py [hunk 9]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+    badge_names = [b.name for b in result["badges"]]
+    for name in must_contain:
+        assert name in badge_names
+
+
+@pytest.mark.parametrize(
+    "query,begin,limit,expected_begin,expected_limit,max_results",
```
Comment (sdglitched) [inline tests/test_dbapi.py]: `max_results` will always be the `limit`, so testing that is not fruitful. Test what the expected count would be, and if the total matches or exceeds it
```suggestion
    "query, begin, limit, expected_begin, expected_limit, expected_count",
```

File: tests/test_dbapi.py [hunk 10]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+    [
+        ("python", 0, 100, 0, 100, None),
+        ("python", 0, 1, 0, 1, 1),
+        ("python", 1, 1, 1, 1, 1),
+        ("python", 500, 1, 500, 1, 0),
+        ("", 0, 500, 0, 100, None),
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Make necessary changes in the parameter as per the `expected_count`.

File: tests/test_dbapi.py [hunk 11]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+    ],
+)
+def test_search_pagination_conditions(
+    query, begin, limit, expected_begin, expected_limit, max_results, api, search_test_badges
+):
+    result = api.get_badges_by_string(query, begin=begin, limit=limit)
+
+    assert result["begin"] == expected_begin
+    assert result["limit"] == expected_limit
+    assert result["total"] >= len(result["badges"])
```
Comment (sdglitched) [inline tests/test_dbapi.py]:
```suggestion
    assert len(result["badges"]) == expected_count
    assert result["total"] >= expected_count
```

File: tests/test_dbapi.py [hunk 12]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
...
+
+    if max_results is not None:
+        assert len(result["badges"]) <= max_results
```
Comment (sdglitched) [inline tests/test_dbapi.py]: No need for this check if the above is implemented.

File: tests/test_dbapi.py [hunk 13]
```diff
@@ -675,3 +720,47 @@ def test_get_all_series_empty(api):
     series_query = api.get_all_series()
     series_list = list(series_query)
     assert len(series_list) == 0
+
+
+@pytest.mark.parametrize(
+    "query,expected_total,must_contain",
```
Comment (sdglitched) [inline tests/test_dbapi.py]:
```suggestion
    "query, expected_total, must_contain",
```

**Review by sdglitched — CHANGES_REQUESTED** [review]
@Daniel-1600, good work on this.

Please check the comments for further changes, and we should be good to go with the merging.

**Comment by Daniel-1600** [comment]
> > for the explanation won't it add up the charaters or its is not considered?
> 
> So, the basic commit convention says that the header should have at most 50 characters, and the description is allowed to have 72 characters per line, i.e., there can be multiple lines in the description, but each line should not exceed 72 characters. image
> 
> > ooh and also for the commit message should include changes i am making ie if its a feature ?
> 
> It is good to have, not a hard and fast rule.
> 
> You can refer [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) for more details, but keep in mind its examples do not follow the 50/72 suggestion.

It all makes much more sense now—thank you for the clarification. I’ve learned a lot from this, especially about maintaining a clean commit history. I truly appreciate your guidance.

**Comment by Daniel-1600** [comment]
> @Daniel-1600, good work on this.
> 
> Please check the comments for further changes, and we should be good to go with the merging.

I’ve made all the requested changes based on your feedback. Please let me know if everything looks good for merging.

File: tests/test_dbapi.py [hunk 14]
```diff
@@ -675,3 +720,55 @@ def test_get_all_series_empty(api):
     series_query = api.get_all_series()
     series_list = list(series_query)
     assert len(series_list) == 0
+
+
+@pytest.mark.parametrize(
+    "query, expected_total,must_contain",
```
Comment (sdglitched) [inline tests/test_dbapi.py]:
```suggestion
    "query, expected_total, must_contain",
```

File: tests/test_dbapi.py [hunk 15]
```diff
@@ -675,3 +720,55 @@ def test_get_all_series_empty(api):
...
+    result = api.get_badges_by_string(query)
+
+    assert result["total"] == expected_total
+    badge_names = [b.name for b in result["badges"]]
+    for name in must_contain:
+        assert name in badge_names
+
+
+@pytest.mark.parametrize(
+    "query,begin,limit,expected_begin,expected_limit, expected_count",
```
Comment (sdglitched) [inline tests/test_dbapi.py]:
```suggestion
    "query, begin, limit, expected_begin, expected_limit, expected_count",
```

File: tests/test_dbapi.py [hunk 16]
```diff
@@ -675,3 +720,55 @@ def test_get_all_series_empty(api):
...
+        ("", 0, 500, 0, 100, 5),
+    ],
+)
+def test_search_pagination_conditions(
+    query, begin, limit, expected_begin, expected_count, api, search_test_badges
+):
+    result = api.get_badges_by_string(query, begin=begin, limit=limit)
+
+    assert result["begin"] == expected_begin
+    assert result["limit"] == expected_count
```
Comment (sdglitched) [inline tests/test_dbapi.py]: This condition will not always be true. I'll let you explain it to me.

**Review by sdglitched — CHANGES_REQUESTED** [review]
The workflows are failing. Please fix them, and along with that, make the suggested changes.

File: tests/test_dbapi.py (hunk 16 above)
Comment (Daniel-1600) [reply tests/test_dbapi.py]: result["limit"] is just the page size echoed back — it's whatever I passed in as limit. expected_count is how many badges actually match the query. They're equal in my test only because I happened to size the fixture data to match the limit. If there were fewer results than the limit, the assertion would break. Should be len(result["badges"]) == expected_count instead.

Comment (sdglitched) [reply tests/test_dbapi.py]: Good explanation👍

File: tests/test_dbapi.py [hunk 17]
```diff
@@ -675,3 +720,55 @@ def test_get_all_series_empty(api):
...
+        ("python", 500, 1, 500, 1, 0),
+        ("", 0, 500, 0, 100, 5),
+    ],
+)
+def test_search_pagination_conditions(
+    query, begin, limit, expected_begin, expected_limit, expected_count, api, search_test_badges
+):
+    result = api.get_badges_by_string(query, begin=begin, limit=limit)
+
+    assert result["begin"] == expected_begin
```
Comment (sdglitched) [inline tests/test_dbapi.py]: Just a minute nitpick: `result["begin"]` isn't being changed in the function you created. So, testing this is futile. Please remove it.

**Review by sdglitched — CHANGES_REQUESTED** [review]
This probably will be the last change before we can merge it.

**Comment by Daniel-1600** [comment]
> This probably will be the last change before we can merge it.

on it lemme make the changes

**Review by sdglitched — APPROVED** [review]
LGTM 🚀

@Daniel-1600, thank you for working on this.

@gridhead, we can merge this PR.

**Comment by sdglitched** [comment]
@gridhead, this PR is ready for merging. Once you are around and available, please take a look.

**Review by gridhead — CHANGES_REQUESTED** [review]
@Daniel-1600, since #278 was merged, there were some merge conflicts introduced here. Please resolve them and we should be able to merge this pull request too.

**Comment by Daniel-1600** [comment]
@gridhead I've resolved the merge conflicts.It's ready for your review

**Review by sdglitched — CHANGES_REQUESTED** [review]
@Daniel-1600, it seems you have not resolved the merge conflict correctly.

The test functions `test_get_persons_by_nickname_search` and `test_get_persons_by_nickname_pagination` are being removed by this PR's changes, which should not be the case.

Please correct this so that no tests are removed.

**Review by sdglitched — APPROVED** [review]
LGTM 🚀 

@gridhead, this PR can now be merged.

**Review by gridhead — APPROVED** [review]
LGTM 🎉
